### PR TITLE
Add `Migrate to IFS` action

### PR DESCRIPTION
**Pull request by SanjulaGanepola**
### Changes
- Add right-click `Migrate to IFS` action on libraries which will migrate the members in select object files to the deploy location in IFS

### CODE REVIEW

**Comment by worksofliam** [comment]
I don't think migrating the IFS is enough here. I also think after moving the source to the IFS, we should automating downloading it to a directory on the users device (there is a `downloadFolder` API), clearing up the data on the server and then asking the user if they want to open the downloaded directory in VS Code.

Migrating the IFS just isn't enough, we should ideally be helping them every step of the way.

**Review by worksofliam — CHANGES_REQUESTED** [review]
I don't think migrating the IFS is enough here. I also think after moving the source to the IFS, we should automating downloading it to a directory on the users device (there is a downloadFolder API), clearing up the data on the server and then asking the user if they want to open the downloaded directory in VS Code.

Migrating the IFS just isn't enough and we should ideally be helping them every step of the way.

File: src/views/projectExplorer/index.ts [hunk 1]
```diff
@@ -54,6 +56,66 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
...
+            };
+            await window.withProgress({
+              location: ProgressLocation.Notification,
+              title: l10n.t('Migrating to IFS'),
+            }, async (progress) => {
+              for await (const file of migrationData.sourceFiles) {
+                progress.report({ message: file });
+
+                const result = await ibmi?.getConnection().sendCommand({
+                  command: `export PATH="/QOpenSys/pkgs/bin:$PATH:" && /QOpenSys/pkgs/bin/makei cvtsrcpf -c ${migrationData.defaultCCSID} ${path.parse(file).name} ${element.label}`,
```
Comment (worksofliam) [inline src/views/projectExplorer/index.ts:90]: We likely need to add a check before this runs to ensure `/QOpenSys/pkgs/bin/makei` is installed/exists.

File: src/views/projectExplorer/index.ts [hunk 2]
```diff
@@ -54,6 +56,66 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
...
+                  directory: remoteDir
+                });
+
+                if (result?.code === 0) {
+                  migration.numSuccess++;
```
Comment (worksofliam) [inline src/views/projectExplorer/index.ts:95]: Consider a `progress.report` here to show to the user that there is progress in the migration.

File: src/views/projectExplorer/migrateToIFS.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { IBMiFile } from "@halcyontech/vscode-ibmi-types";
+
+export async function migrateToIFS(library: string) {
+    let sourceFiles: IBMiFile[] | undefined;
+    await window.withProgress({
+        location: { viewId: `projectExplorer` },
+        title: l10n.t('Migrating to IFS'),
+    }, async () => {
+        const ibmi = getInstance();
+        sourceFiles = await ibmi?.getContent().getObjectList({ library: library });
```
Comment (worksofliam) [inline src/views/projectExplorer/migrateToIFS.ts:16]: The parameter for `getObjectList` can be this to just fetch source files: 

```
{ library: library, types: '*SRCPF' }
```

**Review by worksofliam — CHANGES_REQUESTED** [review]
Some additional code changes which could be made.

File: src/views/projectExplorer/index.ts (hunk 1 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/index.ts]: Good point, added this check

File: src/views/projectExplorer/index.ts (hunk 2 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/index.ts]: Updated the progress to show the source file being migrated and also when downloading the files to the workspace folder

**Comment by SanjulaGanepola** [comment]
> I don't think migrating the IFS is enough here. I also think after moving the source to the IFS, we should automating downloading it to a directory on the users device (there is a downloadFolder API), clearing up the data on the server and then asking the user if they want to open the downloaded directory in VS Code.
> 
> Migrating the IFS just isn't enough and we should ideally be helping them every step of the way.

Updated the `CustomUI` to include the deploy location and the workspace folder to which the files will be migrated and then downloaded to. The default workspace folder will be the one associated with the project. There is also an option to `Add Folder to Workspace` where the user can start a new project. This folder will also automatically be added to the workspace.

image

**Review by edmundreinhardt — APPROVED** [review]

File: src/views/projectExplorer/migrateSource.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        if (isCvtsrcpfInstalled?.code !== 0 || isCvtsrcpfInstalled.stdout !== '/QOpenSys/pkgs/bin/makei') {
+            window.showErrorMessage(l10n.t('Required component ({0}) to migrate source not installed on host IBM i', '/QOpenSys/pkgs/bin/makei'));
+            return;
+        }
+
+        // Create remote directory if it does not exist
+        const mkdirResult = await ibmi?.getConnection().sendCommand({
+            command: `mkdir -p ${migrationConfig.deployLocation}`
+        });
+        if (mkdirResult && mkdirResult.code !== 0) {
```
[the remark this answers is not in the document]
Comment (SanjulaGanepola) [reply src/views/projectExplorer/migrateSource.ts:50]: The `-p` options in `mkdir` will ignore the error if it exists

**Review by edmundreinhardt — APPROVED** [review]

**Pull request merged**